src/job.rs

```diff
@@ -0,0 +1,247 @@
+use diesel::dsl::now;
+use diesel::prelude::*;
+use diesel::{insert_into, update};
+use diesel::sql_types::Integer;
+use serde::{Serialize, Deserialize};
+use serde_json;
+use std::collections::HashMap;
+
+use schema::background_jobs;
+use util::errors::*;
+
+#[derive(Queryable, Identifiable, Debug, Clone)]
+pub struct BackgroundJob {
+    id: i64,
+    job_type: String,
+    data: serde_json::Value,
+}
+
+pub type JobRegistry = HashMap<&'static str, fn(&PgConnection, BackgroundJob) -> CargoResult<()>>;
+
+pub trait Job: Serialize + for<'a> Deserialize<'a> {
+    const JOB_TYPE: &'static str;
+
+    fn enqueue(self, conn: &PgConnection) -> CargoResult<()> {
+        use schema::background_jobs::dsl::*;
+
+        let job_data = serde_json::to_value(self)?;
+        insert_into(background_jobs)
+            .values((
+                job_type.eq(Self::JOB_TYPE),
+                data.eq(job_data),
+            ))
+            .execute(conn)?;
+        Ok(())
+    }
+
+    fn try_perform(conn: &PgConnection, job: BackgroundJob) -> CargoResult<()> {
+        let data = serde_json::from_value(job.data)?;
+        Self::perform(data, conn)?;
+        Ok(())
+    }
+
+    fn register(registry: &mut JobRegistry) {
+        registry.insert(Self::JOB_TYPE, Self::try_perform);
+    }
+
+    fn perform(self, conn: &PgConnection) -> CargoResult<()>;
+}
+
+sql_function!(power, power_t, (x: Integer, y: Integer) -> Integer);
+
+fn get_single_job<F>(conn: &PgConnection, f: F) -> CargoResult<()>
+where
+    F: FnOnce(BackgroundJob) -> CargoResult<()>,
+{
+    use schema::background_jobs::dsl::*;
+    use diesel::dsl::*;
+    use diesel::sql_types::Interval;
+
+    let mut job = None;
+    conn.transaction::<_, Box<dyn CargoError>, _>(|| {
+        let found_job = background_jobs
+            .select((id, job_type, data))
+            .filter(last_retry.lt(now - 1.minute().into_sql::<Interval>() * power(2, retries)))
+            .order(id)
+            .for_update()
+            .skip_locked()
+            .first::<BackgroundJob>(conn)?;
+        increment_retries(conn, found_job.id);
+        job = Some(found_job);
+        Ok(())
+    })?;
+
+    if let Some(job) = job {
+        conn.transaction(|| {
+            delete(&job).execute(conn)?;
```

Comment:
If I understand the locking strategy here, the first transaction above ensures that no 2 threads will fetch the same job.  Once a job is fetched and the retry info is updated, the query will not fetch that row again for at least 1 minute.

Very shortly after the first transaction is completed (and that row lock released), the delete statement in this transaction will again lock the row for update such that even a job that runs for longer than a minute will not be fetched again by another thread until the row is unlocked.

Is my understanding here correct?  Is there an advantage/requirement to do these steps in 2 separate transactions?  I ask because it took me a little while to realize that this delete is re-locking the row and that it is important to issue this delete within the transaction *before* executing the job in the next line to ensure the row is locked while the job runs.

If somehow it took more than a minute for a thread to progress from line 72 to 76, and another thread fetched the same job, then I believe one of them would block on this delete (or maybe abort due to a detected deadlock).  If the other transaction completed successfully (deleting the row), then this would return an error.  If the other transaction failed, then the row would still exist and be unlocked so the other thread would immediately retry the job.  I don't think this is a reasonable failure mode, but I think a single execution of the job would still be guaranteed under this extreme scenario.

Reply:
> Is my understanding here correct?

Yes

> ? Is there an advantage/requirement to do these steps in 2 separate transactions? 

It's a requirement. We cannot update the retry counter in the same transaction as the job itself, since if the job fails the transaction may be aborted meaning we cannot commit those changes.

> If somehow it took more than a minute for a thread to progress from line 72 to 76

We will hit a TCP timeout before a minute passes. The only way it could take more than a minute to get from line 72 to 76 is if the OS has actually pre-empted our process and we receive no CPU resources for a full minute, which is not realistic.

But yes, you're correct that this behavior relies on less than a minute passing between the first transaction committing and the second transaction locking the row.

> and another thread fetched the same job, then I believe one of them would block on this delete (or maybe abort due to a detected deadlock).

Yeah, it'd block on the delete. I think you're right that we should add a sanity check to ensure the row was *actually* deleted before running the job as a sanity check. At the point where the `delete` line returns, we can be sure that we have an exclusive lock on the row, so that would be the point to make sure "no seriously we are the only thing that has access to this 100%". It's not a realistic failure scenario, but it doesn't hurt to handle it.

Reply:
> It's a requirement. We cannot update the retry counter in the same transaction as the job itself, since if the job fails the transaction may be aborted meaning we cannot commit those changes.

Ah, great point!  I knew I was missing some subtlety here.

Overall this looks really good to me.  While we'll only need one consumer for git operations, this is well future proofed for additional use cases.

Reply:
Yes, I've written this code to be extracted into its own library at some point. The only thing that should be coupled to crates.io is the use of `CargoResult`. To make this generically applicable to us I will specifically need to add a way to have a differing number of workers for different job types (having >1 worker for the git job makes no sense), but we don't have to deal with that for now since the git job will be our only job